internal/common/expand_into_runfiles.bzl

```diff
@@ -15,47 +15,91 @@
-    path = path.replace(ctx.genfiles_dir.path + "/", "", 1)
-    return path
-
 def expand_location_into_runfiles(ctx, input, targets = []):
-    """Expands all $(location ...) templates in the given string by replacing $(location //x) with the path
-    in runfiles of the output file of target //x. Expansion only works for labels that point to direct dependencies
+    """Expands all $(execpath ...), $(rootpath ...) and legacy $(location ...) templates in the
+    given string by replacing with the expanded path. Expansion only works for labels that point to direct dependencies
     of this rule or that are explicitly listed in the optional argument targets.
 
-    Path is returned in runfiles manifest path format such as `repo/path/to/file`. This differs from how $(location)
-    and $(locations) expansion behaves in expansion the `args` attribute of a *_binary or *_test which returns
-    the runfiles short path of the format `./path/to/file` for user repo and `../external_repo/path/to/file` for external
-    repositories. We may change this behavior in the future with $(mlocation) and $(mlocations) used to expand
-    to the runfiles manifest path.
+    See https://docs.bazel.build/versions/master/be/make-variables.html#predefined_label_variables.
+
+    Use $(rootpath) and $(rootpaths) to expand labels to the runfiles path that a built binary can use
+    to find its dependencies. This path is of the format:
+    - `./file`
+    - `path/to/file`
+    - `../external_repo/path/to/file`
+
+    Use $(execpath) and $(execpaths) to expand labels to the execroot (where Bazel runs build actions).
+    This is of the format:
+    - `./file`
+    - `path/to/file`
+    - `external/external_repo/path/to/file`
+    - `<bin_dir>/path/to/file`
+    - `<bin_dir>/external/external_repo/path/to/file`
+
+    The legacy $(location) and $(locations) expansion is DEPRECATED as it returns the runfiles manifest path of the
+    format `repo/path/to/file` which behaves differently than the built-in $(location) expansion in args of *_binary
+    and *_test rules which returns the rootpath.
     See https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-binaries.
+    This also differs from how the builtin ctx.expand_location() expansions of $(location) and ($locations) behave
+    as that function returns either the execpath or rootpath depending on the context.
+    See https://docs.bazel.build/versions/master/be/make-variables.html#predefined_label_variables.
+
+    The behavior of $(location) and $(locations) expansion may change in the future with support either being removed
+    entirely or the expansion changed to return the same path as ctx.expand_location() returns for these.
+
+    The recommended approach is to now use $(rootpath) where you previously used $(location). To get from a $(rootpath)
+    to the absolute path that $$(rlocation $(location)) returned you can either use $$(rlocation $(rootpath)) if you
```

Comment:
it would be nice to explain this one level deeper - what are the two dollar signs doing, and does one expansion happen before the action and the other happens inside it? Do we describe the rlocation helper itself somewhere?

Reply:
👍 I'll add more details to the comment.

Since this adds expansion of predefined & custom "make" variables (https://docs.bazel.build/versions/master/be/make-variables.html) such as $(COMPILATION_MODE) $(BINDIR) & $(TARGET_CPU) using `ctx.expand_make_variables`:

```
    # First expand predefined source/output path variables:
    # $(execpath), $(rootpath) & legacy $(location)
    expanded_args = [expand_location_into_runfiles(ctx, a, ctx.attr.data) for a in ctx.attr.templated_args]

    # Next expand predefined variables & custom variables
    expanded_args = [ctx.expand_make_variables("templated_args", e, {}) for e in expanded_args]
```

Bazel will error out if it sees the single dollar sign `$(rlocation path)`. For expand_make_variables not to expand $(rlocation) which is meant be be left in the arg and expanded in the launcher.sh we need to now use `$$(rlocation)` which expands to `$(rlocation)` so that `$$(rlocation $(rootpath //:some_file))` expands to `$(rlocation ./some_file)` and bash runs the `rlocation` runfiles.bash helper function to convert`./some_file` to its absolute path.

Reply:
Added some more details